src/cmd/unit_generic.cpp

```diff
@@ -6430,21 +6430,48 @@ bool Unit::UpgradeMounts( const Unit *up,
                                 int tmpammo = mounts[jmod].ammo;
                                 if (mounts[jmod].ammo != -1 && up->mounts[i].ammo != -1) {
                                     tmpammo += up->mounts[i].ammo;
-                                    if (templ) {
-                                        if (templ->GetNumMounts() > jmod) {
-                                            if (templ->mounts[jmod].volume != -1) {
-                                                if (templ->mounts[jmod].volume < mounts[jmod].type->volume*tmpammo) {
-                                                    tmpammo =
-                                                        (int) floor( .125
-                                                                    +( (0
-                                                                        +templ->mounts[jmod].volume)
-                                                                      /mounts[jmod].type->volume ) );
+                                    if (ismissiletype) {
+                                        if (templ) {
+                                            if (templ->GetNumMounts() > jmod) {
+                                                if (templ->mounts[jmod].volume != -1) {
+                                                    if (templ->mounts[jmod].volume < mounts[jmod].type->volume*tmpammo) {
+                                                        tmpammo =
+                                                            (int) floor( .125
+                                                                        +( (0
+                                                                            +templ->mounts[jmod].volume)
+                                                                          /mounts[jmod].type->volume ) );
+                                                    }
                                                 }
                                             }
                                         }
+                                        if (tmpammo*mounts[jmod].type->volume > mounts[jmod].volume)
+                                            tmpammo = (int) floor( .125+( (0+mounts[jmod].volume)/mounts[jmod].type->volume ) );
+                                    } else {
+                                        std::string ammoname = up->name.get();
+                                        std::size_t ammopos = ammoname.find("_ammo");
+                                        std::string weaponname = ammoname.substr(0, ammopos);
+
+                                        /* Do NOT delete this Unit because it will be either fetched
+                                        * fetched from a cache or - if it has to be created - it will
+                                        * be automatically put in a cache.
+                                        * Deletion will corrupt the cache!
+                                        */
```

Comment:
Also: Will it introduce a memory leak if this unit never gets deleted? That wouldn't be good either.

I am working on a major code overhaul that would use smart pointers wherever possible, so that we wouldn't have to worry about when to manually delete objects. See issue #8 for further discussion. (The code itself isn't ready for review yet.)

Reply:
No, I do not think it will. Here is the code for getUnitFromUpgradeName() (file: src/cmd/unit_functions_generic.cpp):

```
const Unit * getUnitFromUpgradeName( const string &upgradeName, int myUnitFaction = 0 )
{
    const char *name     = upgradeName.c_str();
    const Unit *partUnit = UnitConstCache::getCachedConst( StringIntKey( name, FactionUtil::GetUpgradeFaction() ) );
    if (!partUnit) {
        partUnit = UnitConstCache::setCachedConst( StringIntKey( name,
                                                                FactionUtil::GetUpgradeFaction() ),
                                                  UnitFactory::createUnit( name, true, FactionUtil::GetUpgradeFaction() ) );
    }
    if (partUnit->name == "LOAD_FAILED") {
        partUnit = UnitConstCache::getCachedConst( StringIntKey( name, myUnitFaction ) );
        if (!partUnit)
            partUnit = UnitConstCache::setCachedConst( StringIntKey( name, myUnitFaction ),
                                                      UnitFactory::createUnit( name, true, myUnitFaction ) );
    }
    return partUnit;
}
```

As you can see, it always tries to fetch the requested unit from the cache before it creates it. I verified by inserting logging statements when the function gets called. It does not "forget" the previously created unit and does not start filling the memory with redundant copies. In fact, when I tested it, it never created a new unit, all units were already in the cache. I also checked other usages of the method in the code. They all just call "getUnitFromUpgradeName()". No further postprocessing is required.

Reply:
Good to know. It sounds, then, like there is confusion between different parts of the code over what constitutes a "Unit": Is it a unit **type**, with associated Mesh and other attributes, as loaded from a data file? Or is it an **instance** of said unit type?

It sounds like your code change here is correct, though.